In [1]:
import pandas as pd
import numpy as np



import warnings
warnings.filterwarnings('ignore')

calendar= pd.read_csv(r'C:\Users\Varun\Desktop\Project\Dataset\m5_best_data\m5-forecasting-accuracy\calendar.csv')
price=  pd.read_csv(r'C:\Users\Varun\Desktop\Project\Dataset\m5_best_data\m5-forecasting-accuracy\sell_prices.csv')
sales=  pd.read_csv(r'C:\Users\Varun\Desktop\Project\Dataset\m5_best_data\m5-forecasting-accuracy\sales_train_evaluation.csv')

In [2]:
sales

id        item_id    dept_id   cat_id  \
0      HOBBIES_1_001_CA_1_evaluation  HOBBIES_1_001  HOBBIES_1  HOBBIES   
1      HOBBIES_1_002_CA_1_evaluation  HOBBIES_1_002  HOBBIES_1  HOBBIES   
2      HOBBIES_1_003_CA_1_evaluation  HOBBIES_1_003  HOBBIES_1  HOBBIES   
3      HOBBIES_1_004_CA_1_evaluation  HOBBIES_1_004  HOBBIES_1  HOBBIES   
4      HOBBIES_1_005_CA_1_evaluation  HOBBIES_1_005  HOBBIES_1  HOBBIES   
...                              ...            ...        ...      ...   
30485    FOODS_3_823_WI_3_evaluation    FOODS_3_823    FOODS_3    FOODS   
30486    FOODS_3_824_WI_3_evaluation    FOODS_3_824    FOODS_3    FOODS   
30487    FOODS_3_825_WI_3_evaluation    FOODS_3_825    FOODS_3    FOODS   
30488    FOODS_3_826_WI_3_evaluation    FOODS_3_826    FOODS_3    FOODS   
30489    FOODS_3_827_WI_3_evaluation    FOODS_3_827    FOODS_3    FOODS   

      store_id state_id  d_1  d_2  d_3  d_4  ...  d_1932  d_1933  d_1934  \
0         CA_1       CA    0    0    0    0  ...       2       4       0   
1         CA_1       CA    0    0    0    0  ...       0       1       2   
2         CA_1       CA    0    0    0    0  ...       1       0       2   
3         CA_1       CA    0    0    0    0  ...       1       1       0   
4         CA_1       CA    0    0    0    0  ...       0       0       0   
...        ...      ...  ...  ...  ...  ...  ...     ...     ...     ...   
30485     WI_3       WI    0    0    2    2  ...       1       0       3   
30486     WI_3       WI    0    0    0    0  ...       0       0       0   
30487     WI_3       WI    0    6    0    2  ...       0       0       1   
30488     WI_3       WI    0    0    0    0  ...       1       1       1   
30489     WI_3       WI    0    0    0    0  ...       1       2       0   

       d_1935  d_1936  d_1937  d_1938  d_1939  d_1940  d_1941  
0           0       0       0       3       3       0       1  
1           1       1       0       0       0       0       0  
2           0       0       0       2       3       0       1  
3           4       0       1       3       0       2       6  
4           2       1       0       0       2       1       0  
...       ...     ...     ...     ...     ...     ...     ...  
30485       0       1       1       0       0       1       1  
30486       0       0       0       1       0       1       0  
30487       2       0       1       0       1       0       2  
30488       4       6       0       1       1       1       0  
30489       5       4       0       2       2       5       1  

[30490 rows x 1947 columns]

In [3]:
sales.id.unique().tolist()

['HOBBIES_1_001_CA_1_evaluation',
 'HOBBIES_1_002_CA_1_evaluation',
 'HOBBIES_1_003_CA_1_evaluation',
 'HOBBIES_1_004_CA_1_evaluation',
 'HOBBIES_1_005_CA_1_evaluation',
 'HOBBIES_1_006_CA_1_evaluation',
 'HOBBIES_1_007_CA_1_evaluation',
 'HOBBIES_1_008_CA_1_evaluation',
 'HOBBIES_1_009_CA_1_evaluation',
 'HOBBIES_1_010_CA_1_evaluation',
 'HOBBIES_1_011_CA_1_evaluation',
 'HOBBIES_1_012_CA_1_evaluation',
 'HOBBIES_1_013_CA_1_evaluation',
 'HOBBIES_1_014_CA_1_evaluation',
 'HOBBIES_1_015_CA_1_evaluation',
 'HOBBIES_1_016_CA_1_evaluation',
 'HOBBIES_1_017_CA_1_evaluation',
 'HOBBIES_1_018_CA_1_evaluation',
 'HOBBIES_1_019_CA_1_evaluation',
 'HOBBIES_1_020_CA_1_evaluation',
 'HOBBIES_1_021_CA_1_evaluation',
 'HOBBIES_1_022_CA_1_evaluation',
 'HOBBIES_1_023_CA_1_evaluation',
 'HOBBIES_1_024_CA_1_evaluation',
 'HOBBIES_1_025_CA_1_evaluation',
 'HOBBIES_1_026_CA_1_evaluation',
 'HOBBIES_1_027_CA_1_evaluation',
 'HOBBIES_1_028_CA_1_evaluation',
 'HOBBIES_1_029_CA_1_evaluation',
 'HOBBIES_1_03

In [4]:
price

store_id        item_id  wm_yr_wk  sell_price
0           CA_1  HOBBIES_1_001     11325        9.58
1           CA_1  HOBBIES_1_001     11326        9.58
2           CA_1  HOBBIES_1_001     11327        8.26
3           CA_1  HOBBIES_1_001     11328        8.26
4           CA_1  HOBBIES_1_001     11329        8.26
...          ...            ...       ...         ...
6841116     WI_3    FOODS_3_827     11617        1.00
6841117     WI_3    FOODS_3_827     11618        1.00
6841118     WI_3    FOODS_3_827     11619        1.00
6841119     WI_3    FOODS_3_827     11620        1.00
6841120     WI_3    FOODS_3_827     11621        1.00

[6841121 rows x 4 columns]

In [5]:
price.sort_values(by=['item_id','store_id'])

store_id          item_id  wm_yr_wk  sell_price
368746      CA_1      FOODS_1_001     11101        2.00
368747      CA_1      FOODS_1_001     11102        2.00
368748      CA_1      FOODS_1_001     11103        2.00
368749      CA_1      FOODS_1_001     11104        2.00
368750      CA_1      FOODS_1_001     11105        2.00
...          ...              ...       ...         ...
6512550     WI_3  HOUSEHOLD_2_516     11617        5.94
6512551     WI_3  HOUSEHOLD_2_516     11618        5.94
6512552     WI_3  HOUSEHOLD_2_516     11619        5.94
6512553     WI_3  HOUSEHOLD_2_516     11620        5.94
6512554     WI_3  HOUSEHOLD_2_516     11621        5.94

[6841121 rows x 4 columns]

In [6]:
price.item_id.unique()

array(['HOBBIES_1_001', 'HOBBIES_1_002', 'HOBBIES_1_003', ...,
       'FOODS_3_825', 'FOODS_3_826', 'FOODS_3_827'], dtype=object)

In [7]:
calendar

date  wm_yr_wk    weekday  wday  month  year       d  \
0     2011-01-29     11101   Saturday     1      1  2011     d_1   
1     2011-01-30     11101     Sunday     2      1  2011     d_2   
2     2011-01-31     11101     Monday     3      1  2011     d_3   
3     2011-02-01     11101    Tuesday     4      2  2011     d_4   
4     2011-02-02     11101  Wednesday     5      2  2011     d_5   
...          ...       ...        ...   ...    ...   ...     ...   
1964  2016-06-15     11620  Wednesday     5      6  2016  d_1965   
1965  2016-06-16     11620   Thursday     6      6  2016  d_1966   
1966  2016-06-17     11620     Friday     7      6  2016  d_1967   
1967  2016-06-18     11621   Saturday     1      6  2016  d_1968   
1968  2016-06-19     11621     Sunday     2      6  2016  d_1969   

      event_name_1 event_type_1  event_name_2 event_type_2  snap_CA  snap_TX  \
0              NaN          NaN           NaN          NaN        0        0   
1              NaN          NaN           NaN          NaN        0        0   
2              NaN          NaN           NaN          NaN        0        0   
3              NaN          NaN           NaN          NaN        1        1   
4              NaN          NaN           NaN          NaN        1        0   
...            ...          ...           ...          ...      ...      ...   
1964           NaN          NaN           NaN          NaN        0        1   
1965           NaN          NaN           NaN          NaN        0        0   
1966           NaN          NaN           NaN          NaN        0        0   
1967           NaN          NaN           NaN          NaN        0        0   
1968  NBAFinalsEnd     Sporting  Father's day     Cultural        0        0   

      snap_WI  
0           0  
1           0  
2           0  
3           0  
4           1  
...       ...  
1964        1  
1965        0  
1966        0  
1967        0  
1968        0  

[1969 rows x 14 columns]

In [8]:
calendar.wm_yr_wk.unique()

array([11101, 11102, 11103, 11104, 11105, 11106, 11107, 11108, 11109,
       11110, 11111, 11112, 11113, 11114, 11115, 11116, 11117, 11118,
       11119, 11120, 11121, 11122, 11123, 11124, 11125, 11126, 11127,
       11128, 11129, 11130, 11131, 11132, 11133, 11134, 11135, 11136,
       11137, 11138, 11139, 11140, 11141, 11142, 11143, 11144, 11145,
       11146, 11147, 11148, 11149, 11150, 11151, 11152, 11201, 11202,
       11203, 11204, 11205, 11206, 11207, 11208, 11209, 11210, 11211,
       11212, 11213, 11214, 11215, 11216, 11217, 11218, 11219, 11220,
       11221, 11222, 11223, 11224, 11225, 11226, 11227, 11228, 11229,
       11230, 11231, 11232, 11233, 11234, 11235, 11236, 11237, 11238,
       11239, 11240, 11241, 11242, 11243, 11244, 11245, 11246, 11247,
       11248, 11249, 11250, 11251, 11252, 11301, 11302, 11303, 11304,
       11305, 11306, 11307, 11308, 11309, 11310, 11311, 11312, 11313,
       11314, 11315, 11316, 11317, 11318, 11319, 11320, 11321, 11322,
       11323, 11324,

# Analysing the datasets

## For Train dataset

* Unique id column which had info (catoegory ,cat num and store number)
* item id which is in price and calendar 
* dep_id denotes which department the product is
* store_id denotes which store the product is
* and number of d_... columns from 1 to 1941 days

## price dataset

* store_id same as in train
* item_id same as in train
* wm_yr_wk (wallmart_year_week) this gives the chronological sequence but not ordered
* sell_price gives the price of the product that week

## calendar dataset

* date gives the date of the record
* wm_yr_wk same as in price dataset but it is ordered in chronlogical order
* weekday and wday gives the same thing but we can say wday is label encoded
* month gives what month ranges(1 to 12)
* year gives what year it is
  * These last 3 or 4 columns can be derived from date

* d gives the details about the count of day which is same number in columns of train ranges(1 to 1941)
* event_name and event_type (4 columns) these gives the event and type of event happening in that particular day if there are 2 events in a day thus there are 2 columns of each
* snap is a system in US (Supplemental Nutrition Assistance Program) which the government give money to buy food the elegibility is given to certain stores at certain times in (0,1)


## Additional details

* wm_yr_wk has values from 11101 (1_11_01 stands as wm_yr_wk), since in a year there can be 52 week max the last 2 digits cannot be more than 52
* the id in train is combination of item_id and store_id
* the dep_id in train can be derived from dept_id 

In [9]:
pd.concat([sales,price])

MemoryError: Unable to allocate 98.9 GiB for an array with shape (1941, 6841121) and data type float64

## Here I cant concat the datasets due to memory error....
***


In [10]:
price.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6841121 entries, 0 to 6841120
Data columns (total 4 columns):
 #   Column      Dtype  
---  ------      -----  
 0   store_id    object 
 1   item_id     object 
 2   wm_yr_wk    int64  
 3   sell_price  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 957.5 MB


In [11]:
sales.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30490 entries, 0 to 30489
Columns: 1947 entries, id to d_1941
dtypes: int64(1941), object(6)
memory usage: 463.3 MB


In [12]:
calendar.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1969 entries, 0 to 1968
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          1969 non-null   object
 1   wm_yr_wk      1969 non-null   int64 
 2   weekday       1969 non-null   object
 3   wday          1969 non-null   int64 
 4   month         1969 non-null   int64 
 5   year          1969 non-null   int64 
 6   d             1969 non-null   object
 7   event_name_1  162 non-null    object
 8   event_type_1  162 non-null    object
 9   event_name_2  5 non-null      object
 10  event_type_2  5 non-null      object
 11  snap_CA       1969 non-null   int64 
 12  snap_TX       1969 non-null   int64 
 13  snap_WI       1969 non-null   int64 
dtypes: int64(7), object(7)
memory usage: 737.5 KB


### These are the memory usages

Reducing the memory by changing the datatypes of dataframe

like uint8 or bool these take significantly less storage but ranges should be considered

In [13]:
price.describe()

wm_yr_wk    sell_price
count  6.841121e+06  6.841121e+06
mean   1.138294e+04  4.410952e+00
std    1.486100e+02  3.408814e+00
min    1.110100e+04  1.000000e-02
25%    1.124700e+04  2.180000e+00
50%    1.141100e+04  3.470000e+00
75%    1.151700e+04  5.840000e+00
max    1.162100e+04  1.073200e+02

In [14]:
sales.describe()

d_1           d_2           d_3           d_4           d_5  \
count  30490.000000  30490.000000  30490.000000  30490.000000  30490.000000   
mean       1.070220      1.041292      0.780026      0.833454      0.627944   
std        5.126689      5.365468      3.667454      4.415141      3.379344   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max      360.000000    436.000000    207.000000    323.000000    296.000000   

                d_6           d_7           d_8           d_9          d_10  \
count  30490.000000  30490.000000  30490.000000  30490.000000  30490.000000   
mean       0.958052      0.918662      1.244080      1.073663      0.838701   
std        4.785947      5.059495      6.617729      5.917204      4.206199   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max      314.000000    316.000000    370.000000    385.000000    353.000000   

       ...        d_1932        d_1933        d_1934        d_1935  \
count  ...  30490.000000  30490.000000  30490.000000  30490.000000   
mean   ...      1.445490      1.781174      1.965267      1.389374   
std    ...      3.656824      4.426550      4.706284      3.313292   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      0.000000      1.000000      1.000000      0.000000   
75%    ...      2.000000      2.000000      2.000000      2.000000   
max    ...    143.000000    156.000000    187.000000     98.000000   

             d_1936        d_1937        d_1938        d_1939        d_1940  \
count  30490.000000  30490.000000  30490.000000  30490.000000  30490.000000   
mean       1.271794      1.216661      1.212299      1.395605      1.689669   
std        3.126178      3.000348      2.955910      3.514318      4.089208   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      1.000000   
75%        1.000000      1.000000      1.000000      1.000000      2.000000   
max       90.000000     77.000000     84.000000    110.000000    156.000000   

             d_1941  
count  30490.000000  
mean       1.782158  
std        4.284356  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        2.000000  
max      117.000000  

[8 rows x 1941 columns]

In [15]:
calendar.describe()

wm_yr_wk         wday        month         year      snap_CA  \
count   1969.000000  1969.000000  1969.000000  1969.000000  1969.000000   
mean   11347.086338     3.997461     6.325546  2013.288471     0.330117   
std      155.277043     2.001141     3.416864     1.580198     0.470374   
min    11101.000000     1.000000     1.000000  2011.000000     0.000000   
25%    11219.000000     2.000000     3.000000  2012.000000     0.000000   
50%    11337.000000     4.000000     6.000000  2013.000000     0.000000   
75%    11502.000000     6.000000     9.000000  2015.000000     1.000000   
max    11621.000000     7.000000    12.000000  2016.000000     1.000000   

           snap_TX      snap_WI  
count  1969.000000  1969.000000  
mean      0.330117     0.330117  
std       0.470374     0.470374  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       1.000000     1.000000  
max       1.000000     1.000000